### PR TITLE
Cross-file template imports — schema, helpers, vscode preview, example (#277)

### LINKED ISSUE #277: Cross-file template imports for reusable measurement instruments
## Motivation

Researchers often develop Stagebook syntax for a measurement instrument — a survey, a battery of prompts, a scoring rule — that they want to use across multiple studies. Today there's no way to share these: the entire `templates:` block has to be copy-pasted between treatment files, and any update to the instrument has to be replicated by hand.

We want a way to ship a measurement instrument as a self-contained folder (a treatment file with `templates:` only, plus its required prompt files), drop the folder into a study, and import its templates by reference.

### Mechanism

A new top-level `imports:` field on the treatment file:

```yaml
imports:
  - ./surveys/tipi/tipi.stagebook.yaml
  - ./scoring/partisan-7pt/scoring.stagebook.yaml

templates:
  # ...inline templates...
```

The path is relative to the importing treatment file.

### Pipeline (three new Stagebook helpers; host owns the loading loop)

Imports are resolved + path-rewritten *before* any other Stagebook step. The schema sees fully expanded templates and validates them as today. The new pipeline introduces three helpers in the React-free entry point and one new host responsibility (a loading loop):

```typescript
import {
  parseTreatmentYaml,
  resolveImportPath,
  resolveImports,
  fillTemplates,
  treatmentFileSchema,
} from "stagebook";

// 1. Parse the main file. Returns { parsed, imports } — the import list
//    surfaces here so the host never has to peek into Stagebook's schema
//    to find import declarations.
const { parsed: main, imports: mainImports } =
  parseTreatmentYaml(mainYamlString);

// 2. Host loop: load every (transitively) imported file. The host owns
//    sync vs async, error handling, and any extra canonicalization
//    (symlinks, case folding) it wants. Stagebook owns syntactic path
//    normalization via resolveImportPath.
const loaded = new Map<string, ParsedTreatmentFile>();
const queue = mainImports.map((p) =>
  resolveImportPath("./main.yaml", p),
);

while (queue.length) {
  const path = queue.shift()!;
  if (loaded.has(path)) continue;                   // dedup
  const text = await loadFile(path);                // host's loader
  const { parsed, imports } = parseTreatmentYaml(text);
  loaded.set(path, parsed);
  queue.push(...imports.map((p) => resolveImportPath(path, p)));
}

// 3. Stagebook merges + path-rewrites + cycle-detects across the full set.
const templates = resolveImports({ main, files: loaded });

// 4. Existing pipeline runs unchanged.
const hydrated = {
  ...main,
  introSequences: fillTemplates({ obj: main.introSequences, templates }),
  treatments: fillTemplates({ obj: main.treatments, templates }),
};
const result = treatmentFileSchema.safeParse(hydrated);
```

**What each helper owns:**

- `parseTreatmentYaml(yamlString)` → `{ parsed, imports }`. Stagebook bundles a pinned YAML parser (currently `js-yaml`, moved from peer dep to direct runtime dep) with a documented config (e.g., JSON-schema-only — no YAML 1.1 implicit booleans, no Norway problem). The `imports` field on the return value is the canonical place where Stagebook surfaces import declarations; hosts never peek into the schema themselves.
- `resolveImportPath(parentPath, importPath)` → string. Standard `path.posix.normalize`-style canonicalization. Collapses `..` and `.`. Two routes to the same file (e.g. `B/b.stagebook.yaml` from main, `../B/b.stagebook.yaml` from A) canonicalize to the same string, so the host's dedup map works correctly.
- `resolveImports({ main, files })` → templates array. Walks every imported template, prepends the import's directory prefix to every `file:` field, merges the rewritten templates into a flat array. Detects cycles in the dependency graph and rejects them. After this step, imported templates are indistinguishable from inline templates and the existing `fillTemplates` / `treatmentFileSchema` pipeline runs unchanged.

**Format flexibility:** `parseTreatmentYaml` is the recommended path; hosts that want JSON, TOML, or DB-backed treatments call their own parser and provide already-parsed objects to `resolveImports` directly. The host loop accommodates both.

### Path rewriting (the non-obvious detail)

File references inside an imported template are written relative to *the imported file's location*. `resolveImports` prepends the import's directory prefix to every `file:` field. The rewrite is **per template, based on the file the template was declared in** — regardless of which other files transitively imported it. So a template from `surveys/tipi/tipi.stagebook.yaml` has its `file:` paths prefixed with `surveys/tipi/`, producing paths relative to the *main* treatment file's location, no matter which import chain reached it.

Example: importer at `study/main.stagebook.yaml` imports `./surveys/tipi/tipi.stagebook.yaml`. A template inside that file with `file: q1.prompt.md` becomes `file: surveys/tipi/q1.prompt.md` after path rewriting. The host then resolves it relative to the consumer's treatment file, exactly as it does for any inline template today.

### Nested imports, deduplication, and cycles

- **Nested imports are supported.** A measurement-instrument module may itself import shared sub-modules (e.g., a `partisan` module imports a shared `likert-rating` template). The host's load loop walks transitively until the queue drains.
- **Deduplication.** The host's dedup map keys on the canonical path returned by `resolveImportPath`. A file imported by multiple paths (e.g., `B/b.stagebook.yaml` from main and `../B/b.stagebook.yaml` from A both normalize to `B/b.stagebook.yaml`) is loaded exactly once.
- **Symlinks and case sensitivity** are out of Stagebook's scope. Stagebook's normalization is syntactic and case-preserving. Hosts whose filesystems require further canonicalization (resolving symlinks, case folding on Windows) layer it before consulting their dedup map.
- **Cycles are rejected at validation.** `resolveImports` walks the dependency graph it sees across all loaded files; cycles are reported as schema-level errors before any participant interaction.

### Namespacing

Convention-based, not mechanical. Module authors prefix their template and element names with the module's name (`tipi_questions`, `tipi_score`, internal `name: tipi_q1` … `tipi_q10`). Internal references inside the module use the prefixed names; external references use the same prefixed names. No schema-level "module" concept needed.

For studies that need the same module twice (pre/post TIPI), the convention extends via the existing `fields:` substitution. Module author writes `name: ${prefix}_q1`, exposes `prefix` as a template field, consumer invokes:

```yaml
- template: tipi_questions
  fields: { prefix: "preTIPI" }
- template: tipi_questions
  fields: { prefix: "postTIPI" }
```

No new mechanism — purely a documented pattern.

#### Transitive prefix extension (when modules depend on other modules)

The same convention composes recursively. If module A depends on module B (e.g., TIPI uses a shared Likert-rating template internally), A is responsible for **extending its incoming prefix** when delegating to B — not passing it through unchanged:

```yaml
# A/a.stagebook.yaml — depends on B
templates:
  - name: tipi_questions
    contentType: elements
    content:
      - type: prompt
        name: ${prefix}_q1                         # A's own leaf
      - template: likert_scale                     # delegating to B
        fields: { prefix: "${prefix}_likert" }     # extended, not passed through
```

(`fields:` lives on the *invocation* — `{ template, fields }` — not on the template definition. Required fields are implicit from the `${...}` placeholders that appear in `content`. The substitution is recursive: when A's invocation of B contains `prefix: "${prefix}_likert"`, the inner `${prefix}` resolves against A's surrounding scope before being passed down to B.)

When `main.yaml` invokes A twice (pre/post) and also uses B directly, every leaf storage key is unique:

```
main → tipi_questions { prefix: "preTIPI"  } → preTIPI_q1,  preTIPI_likert_response
main → tipi_questions { prefix: "postTIPI" } → postTIPI_q1, postTIPI_likert_response
main → likert_scale   { prefix: "imgRating" } → imgRating_response
```

This relies on `fillTemplates` recursively expanding `${...}` references inside field values (so that `${prefix}_likert` in A's invocation of B is resolved against A's field scope before being passed to B). That recursive expansion is an existing implementation detail; this issue elevates it to a stable contract.

#### Catching violations

If a module author forgets to extend the prefix (passes it through unchanged), B's elements collide between two invocations of A. The storage-key collision warnings (#149 / its successor) catch this at validation time, before participants run. The convention is enforced by the warning, not by the schema.

### Validation order

Today the host validates the treatment file twice: once **before** `fillTemplates` (so authors get errors that reference the template-as-written) and once **after** (as a backstop for hydration bugs). The schema accepts both expanded and unexpanded shapes via `altTemplateContext` ([packages/stagebook/src/schemas/treatment.ts:514-546](packages/stagebook/src/schemas/treatment.ts#L514-L546)), so the same `treatmentFileSchema` runs in both passes.

Imports run **before both validators**. The host strips `imports:` from the parsed object after `resolveImports` returns the merged templates, so the schema never has to know about `imports:` — it stays untouched. Pipeline:

1. Host parses main YAML.
2. (Optional) lightweight import-machinery validator — paths well-formed, no obvious self-loops. Cheap, gives early feedback before any filesystem work.
3. Host loop: load + parse imported files, dedup, build canonical-path map.
4. `resolveImports` merges templates, rewrites paths, detects graph cycles.
5. Host strips `imports:` from the merged object.
6. Pre-fill validator (today's `treatmentFileSchema`).
7. `fillTemplates`.
8. Post-fill validator (today's `treatmentFileSchema`).

The lightweight import validator from step 2 is a separate exported helper (e.g., `validateImportStatements`) — small and useful, but not required for correctness.

### File schema: one shape, one extension

There is no separate "module file" type. All Stagebook files are `*.stagebook.yaml` (per #310) and validate against a single schema where `treatments:`, `introSequences:`, `templates:`, and `imports:` are each optional top-level keys.

Why unified instead of two file types: any treatment file written today might be imported by another study tomorrow that wants to reuse its templates. Forcing a separate schema/extension for "modules" creates a class boundary that doesn't reflect how researchers actually share work. A file that is the *entry point* for one study can be an *imported source* for another, with no syntactic difference.

`treatmentFileSchema` (the existing schema) gets adjusted: `treatments:` and `introSequences:` become optional. The host enforces "an entry-point file must have at least one treatment" at load time, not at schema time. This is a runtime / loading concern, not a syntactic one — the same file can be a valid module without `treatments:` and a valid entry point with `treatments:`, and the schema doesn't need to know which role it's playing.

This also subsumes #280 (preview standalone template files in the viewer) — the viewer's load path no longer distinguishes module files from treatment files. It just opens a `*.stagebook.yaml` file and renders whatever's there: if the file has `treatments:`, it can preview a participant flow; if not, the preview surfaces the templates themselves (e.g., as a list with click-through to inspect each one).

## Decisions needed before implementation

- [ ] Exact field name on the treatment file: `imports:` vs. `templateImports:` vs. other.
- [ ] Import shape: bare path string vs. object form (`{ from: "...", as: "..." }`). Object form leaves room for per-import options later (selective imports, version pins) but the v1 use case never needs them. Recommend bare strings.
- [ ] Behavior on duplicate template names across imports + inline templates. Recommend error at validation time; author renames via the convention.
- [ ] Whether the imported file can itself have `imports:` (nested imports). Recommend yes, with cycle detection at the load step.
- [ ] Whether the imported file can export elements/stages directly or templates only. Recommend templates only for v1.

## Out of scope for v1

- npm-package-style versioned imports (`tipi@v2.0`). Use local file paths; pin a version by copying the folder.
- Selective imports (`imports: [{ from: "./tipi", only: [...] }]`). All-or-nothing.

## Related

- #310 — file extension rename (`*.treatments.yaml` → `*.stagebook.yaml`). This issue's import examples and the unified file-shape decision both depend on the rename. Coordinate so #310 lands first or alongside.

- #280 — preview standalone template files in the viewer (follow-up ergonomics, not blocking).

- #281 — disallow storage-key reuse across phases. The convention-based prefix extension relies on the storage-key collision detector to catch authoring mistakes; #281 widens that detector's scope so module-reuse mistakes (the most natural place for prefix violations) are caught.

- #278 (aggregate values) — a fully-baked measurement instrument ships its scoring logic alongside its prompts. The two designs are independent but compose naturally: a TIPI module would export both `tipi_questions` and `tipi_extraversion_score` templates; the consumer invokes both. v1 of this issue does not require aggregate values to exist, but should not paint itself into a corner that prevents aggregate values from being module-exportable later.

### PR DESCRIPTION
**Pull request by JamesPHoughton**
## Summary

Closes #277. Lets researchers ship a measurement instrument (TIPI questions, partisan-7pt scoring, etc.) as a self-contained Stagebook file that other studies can pull in via:

\`\`\`yaml
imports:
  - ./surveys/tipi/tipi.stagebook.yaml
  - ./scoring/partisan-7pt/scoring.stagebook.yaml

treatments:
  - name: my_study
    gameStages:
      - elements:
          - template: tipi_questions    # defined in surveys/tipi/...
\`\`\`

## What's in this PR

**Schema** (\`packages/stagebook/src/schemas/treatment.ts\`)
- New optional \`imports:\` field on \`treatmentFileSchema\`.
- \`treatments:\` and \`introSequences:\` become optional (unified file shape per #277). Same file can be a runtime entry point or an importable template source.

**Helpers** (new \`packages/stagebook/src/imports/\`)
- \`resolveImportPath(parentPath, importPath)\` — POSIX path canonicalization with deduplication. Pure, no I/O.
- \`resolveImports({ main, files })\` — merges templates from main + every loaded import, with per-template path rewriting (\`file:\` and \`captionsFile:\` get prefixed with the import directory). Detects template-name collisions across the main + imports with a migration-hint error.
- \`parseTreatmentYaml(yamlString)\` — returns \`{ parsed, imports }\`.
- 41 new unit tests across the three helpers.

\`js-yaml\` moved from peer dep to direct dep so consumers don't have to bring their own.

**VS Code extension** (\`apps/vscode/src/extension.ts\`)
- \`parseTreatmentForPreview\` is now async and runs the full import loop (read each file via \`vscode.workspace.fs\`, parse, recurse on nested imports, dedup) before merging templates and expanding. Both callers updated to await.
- Researchers can develop with \`imports:\` in the VS Code preview today.

**Standalone viewer** (\`apps/viewer/src/lib/\`)
- New async \`loadAndMergeTreatmentFile\` helper for hosts that load files asynchronously (URL fetch, etc.).
- \`exampleCatalog.ts\` resolves bundled example imports synchronously at build time. Stores both the original \`yaml\` and the post-merge \`mergedYaml\` on each entry. The build-time glob now picks up nested module files; \`isExampleRoot\` filters them out from catalog entries — they're loaded only via their importer's \`imports:\`.

**Example** (\`examples/imports-walkthrough/\`)
- Root file imports \`./modules/consent.stagebook.yaml\` and invokes its \`shared_consent_step\` template as the first intro step. Demonstrates the import + path-rewrite mechanics end-to-end.

**Docs**
- \`docs/researcher/treatment-files.md\`: new \"Imports\" section.
- \`docs/engineer/integration-guide.md\`: hydration pipeline shows the host-owned loading loop.

## What's deferred (filed as follow-ups)

- **#312** — Standalone viewer URL loader (\`loader.ts\`) wiring. The async helper exists; this just plumbs it into the URL flow.
- **#313** — VS Code \"Preview Expanded Templates\" virtual document. The preview command supports imports; the expand command's sync \`provideTextDocumentContent\` needs to become async.

These aren't blockers — researchers can develop with imports in the VS Code preview today; the URL flow and the expand-templates view are convenience features that come along after.

## Test plan

- [x] 1221 tests pass (945 stagebook + 87 viewer + 189 vscode)
- [x] 41 new unit tests for the import helpers (path canonicalization edge cases, dedup invariant, path rewriting, name collisions, parser safety)
- [x] New schema tests for the unified file shape (module-only, imports+treatments, empty file, invalid \`imports:\` shapes)
- [x] \`imports-walkthrough\` example tested via \`examples.test.ts\` (parse + expand + every prompt file resolves)
- [x] VS Code extension reinstalled locally; \`Stagebook: Preview Treatment\` on the new example renders the imported consent step as the first intro step
- [x] Lint clean

## ../deliberation-lab integration notes

After merging, hosts that want to support imports should follow the host-loading-loop pattern in [docs/engineer/integration-guide.md](docs/engineer/integration-guide.md#treatment-hydration-pipeline) — the schema accepts \`imports:\` automatically, but resolving them is the host's job.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: packages/stagebook/src/imports/resolveImports.ts [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+    }
+  }
+  return result;
+}
+
+function isAbsolutePath(p: string): boolean {
+  return (
+    p.startsWith("/") || p.startsWith("http://") || p.startsWith("https://")
+  );
+}
```
[the remark this answers is not in the document]
Comment (JamesPHoughton) [reply packages/stagebook/src/imports/resolveImports.ts:158]: Fixed in c1841a9. Replaced the literal-prefix check with a generic scheme-detection regex (`/^[a-zA-Z][a-zA-Z0-9+.-]*:/`) — handles every URI scheme (including `asset://`) case-insensitively. Added test cases for `asset://`, `HTTP://`, and `ASSET://`.

File: apps/vscode/src/extension.ts [hunk 2]
```diff
@@ -22,7 +22,11 @@ import {
 import {
   fillTemplates,
   getReferencedAssets,
+  parseTreatmentYaml as parseStagebookYaml,
+  resolveImportPath,
+  resolveImports,
   treatmentFileSchema,
+  type ParsedFile,
 } from "stagebook";
 import { parse as parseYaml } from "yaml";
```
[the remark this answers is not in the document]
Comment (JamesPHoughton) [reply apps/vscode/src/extension.ts]: Fixed in c1841a9. Removed. Slipped past CI because the vscode workspace currently has `echo 'no linter configured yet'` instead of a real lint step — worth filing as a separate housekeeping issue.

File: apps/vscode/src/extension.ts [hunk 3]
```diff
@@ -591,23 +595,93 @@ function getNonce(): string {
...
+  // Strip `imports:` from the root and replace `templates:` with the
+  // merged set. This is what the schema/runtime expects (no imports
+  // in the post-fill world).
+  const { imports: _imports, templates: _origTemplates, ...rest } = root;
+  void _imports;
+  void _origTemplates;
+  const merged: Record<string, unknown> = { ...rest };
+  if (mergedTemplates.length > 0) {
+    merged.templates = mergedTemplates;
+  }
```
[the remark this answers is not in the document]
Comment (JamesPHoughton) [reply apps/vscode/src/extension.ts]: Fixed in c1841a9. Now tracks `rootHadTemplates = "templates" in root` and re-attaches `templates: mergedTemplates` whenever the root explicitly had the key, even if the merged array is empty. Files that never had `templates:` and have no imports continue to omit the key (so the schema's `optional()` path stays clean).

File: apps/viewer/src/lib/loadAndMergeTreatmentFile.ts [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+
+  // Strip `imports:` from the root and replace `templates:` with the
+  // merged set — what the schema/runtime expects post-import.
+  const { imports: _imports, templates: _origTemplates, ...rest } = root;
+  void _imports;
+  void _origTemplates;
+  const merged: Record<string, unknown> = { ...rest };
+  if (mergedTemplates.length > 0) {
+    merged.templates = mergedTemplates;
+  }
```
[the remark this answers is not in the document]
Comment (JamesPHoughton) [reply apps/viewer/src/lib/loadAndMergeTreatmentFile.ts:71]: Fixed in c1841a9 with the same approach as the vscode-side fix — re-attach when `rootHadTemplates`. The three sites (this one, extension.ts, exampleCatalog.ts) all use the same pattern; could be extracted into a shared helper as a follow-up if a fourth consumer appears.

File: apps/viewer/src/lib/exampleCatalog.ts [hunk 5]
```diff
@@ -25,29 +39,117 @@ export interface ExampleEntry {
...
+  const mergedTemplates = resolveImports({ main: root, files: loaded });
+
+  // Strip `imports:`, attach merged templates, re-serialize as JSON
+  // (which is valid YAML) so the existing string-based parse path
+  // can consume the result without changes.
+  const { imports: _imports, templates: _origTemplates, ...rest } = root;
+  void _imports;
+  void _origTemplates;
+  const merged: Record<string, unknown> = { ...rest };
+  if (mergedTemplates.length > 0) {
```
[the remark this answers is not in the document]
Comment (JamesPHoughton) [reply apps/viewer/src/lib/exampleCatalog.ts]: Fixed in c1841a9. Same `rootHadTemplates` preservation as the other two sites.
